frontend/src/components/dashboard/filter/CriterionFilterForm.tsx

```diff
@@ -0,0 +1,85 @@
+import { FilterCriterion } from ".";
+import React from "react";
+import { match } from "ts-pattern";
+import { AstCriterionType } from "@/data-analyzer/analyze-asts";
+import { NoCriterionFilter } from "../criteria/none";
+import { ConditionCriterionFilter } from "../criteria/condition";
+import { FunctionCallCriterionFilter } from "../criteria/function-call";
+import { StatementCriterionFilter } from "../criteria/statement";
+import { ExpressionCriterionFilter } from "../criteria/expression";
+import { LoopCriterionFilter } from "../criteria/loop";
+import { FunctionDeclarationCriterionFilter } from "../criteria/function-declaration";
+import { TestCriterionFilter } from "../criteria/test";
+import { MetaCriterionType } from "../criteria/meta-criterion-type";
+
+const createProps = <Filter extends FilterCriterion>(
+  filter: Filter,
+  setFilters: (
+    updateFilters: (currentFilters: FilterCriterion[]) => FilterCriterion[],
+  ) => void,
+): {
+  value: Filter;
+  onChange: (newFilter: Filter) => void;
+} => ({
+  value: filter,
+  onChange: (newFilter: Filter) => {
+    setFilters((currentFilters) => {
+      const index = currentFilters.findIndex(
+        (f) => f.criterion === filter.criterion,
+      );
+
+      if (index === -1) {
+        return currentFilters;
```

Comment:
shouldn't we append `newFilter`, in this case?
I might be missing omsehting here 🤔